### PR TITLE
[MSHADE-364] drop duplicates handled by a transformer

### PR DESCRIPTION
**Pull request by rmannibucau**
When a transformer handles a resource then the resource being duplicated is ok since the transformer will merge it so no need to log it (in particular as a warning).

---

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHADE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MSHADE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHADE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/shade/DefaultShader.java [hunk 1]
```diff
@@ -256,6 +256,10 @@ else if ( shadeRequest.isShadeSourcesContent() && name.endsWith( ".java" ) )
 
                     addResource( resources, jos, mappedName, entry.getTime(), in );
                 }
+                else
+                {
+                    duplicates.remove( name, jar );
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/shade/DefaultShader.java:261]: Could there be a test for this?

Comment (rmannibucau) [reply src/main/java/org/apache/maven/plugins/shade/DefaultShader.java:261]: Added https://github.com/apache/maven-shade-plugin/pull/48/files#diff-bc8729c3b7ec6843be39d656cb9b2d40R96

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java [hunk 2]
```diff
@@ -113,16 +83,67 @@ public Logger getChildLogger( final String s )
...
+
+        final DefaultShader shaderWithoutTransformer = new DefaultShader();
+        final MockLogger logWithoutTransformer = new MockLogger();
+        shaderWithoutTransformer.enableLogging( logWithoutTransformer );
+        shadeRequest.setResourceTransformers( Collections.<ResourceTransformer>emptyList() );
+        shaderWithoutTransformer.shade( shadeRequest );
+
+        temporaryFolder.delete();
+
+        assertTrue(logWithTransformer.warnMessages.toString(), logWithTransformer.warnMessages.isEmpty());
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java:142]: Is the purpose of this PR to add another warning to the log?

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java [hunk 3]
```diff
@@ -113,16 +83,67 @@ public Logger getChildLogger( final String s )
...
+        assertTrue(logWithoutTransformer.warnMessages.toString(), logWithoutTransformer.warnMessages.containsAll(
+               Arrays.<String>asList( "j1.jar, j2.jar define 1 overlapping resources:", "- foo.txt" ) ) );
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java:144]: resources --> resource when it's singular

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java (hunk 2 above)
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java:142]: to remove it

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java (hunk 3 above)
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java]: fixed

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java [hunk 4]
```diff
@@ -274,4 +295,50 @@ private static DefaultShader newShader()
         return s;
     }
 
+    private static class MockLogger extends AbstractLogger
+    {
+        private final List<String> debugMessages = new ArrayList<>();
+        private final List<String> warnMessages = new ArrayList<>();
+
+        public MockLogger()
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java]: fixed

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java [hunk 5]
```diff
@@ -274,4 +295,50 @@ private static DefaultShader newShader()
...
+        {
+            super( Logger.LEVEL_INFO, "test" );
+        }
+
+        @Override
+        public void debug( final String s, final Throwable throwable )
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java]: fixed

File: src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java [hunk 6]
```diff
@@ -113,16 +83,67 @@ public Logger getChildLogger( final String s )
...
 
-        final String failureDebugMessage = debugMessages.toString();
-        assertTrue(failureDebugMessage, debugMessages.contains(
+        final String failureDebugMessage = logs.debugMessages.toString();
+        assertTrue(failureDebugMessage, logs.debugMessages.contains(
                 "We have a duplicate META-INF/MANIFEST.MF in src/test/jars/plexus-utils-1.4.1.jar" ));
     }
 
+    public void testOverlappingResourcesAreLoggedExceptATransformerHandlesIt() throws Exception {
+        final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java]: ok

**Comment by rmannibucau** [comment]
CI is blue/green, any blocker to merge?

File: src/main/java/org/apache/maven/plugins/shade/DefaultShader.java [hunk 7]
```diff
@@ -338,11 +342,25 @@ private void logSummaryOfDuplicates( Multimap<Collection<File>, String> overlapp
             final Collection<String> overlaps = new ArrayList<>();
             if ( !classes.isEmpty() )
             {
-                overlaps.add( "classes" );
+                if ( resources.size() == 1 )
+                {
+                    overlaps.add( "classe" );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/DefaultShader.java:347]: redundant e?

Comment (rmannibucau) [reply src/main/java/org/apache/maven/plugins/shade/DefaultShader.java]: I'd say french typo  ;). Thanks for the catch, should be fixed now

Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/shade/DefaultShader.java]: I assumed so...

**Pull request merged**